leasing/serializers/land_use_agreement.py

```diff
@@ -523,7 +534,376 @@ def update(self, instance, validated_data):
+                            valid = False
+                    except ObjectDoesNotExist:
+                        valid = False
+
+        if not valid:
+            raise ValidationError(_("Cannot use an inactive receivable type"))
+
+        return data
+
+
+class LandUseAgreementInvoicePaymentCreateUpdateSerializer(
+    FieldPermissionsSerializerMixin, serializers.ModelSerializer
+):
+    id = serializers.IntegerField(required=False)
+
+    class Meta:
+        model = LandUseAgreementInvoicePayment
+        exclude = ("invoice", "deleted")
+
+
+class LandUseAgreementInvoiceCreateSerializer(
+    UpdateNestedMixin,
+    EnumSupportSerializerMixin,
+    FieldPermissionsSerializerMixin,
+    serializers.ModelSerializer,
+):
+    id = serializers.ReadOnlyField()
+    recipient = InstanceDictPrimaryKeyRelatedField(
+        instance_class=Contact,
+        queryset=Contact.objects.all(),
+        related_serializer=ContactSerializer,
+        required=False,
+    )
+    rows = LandUseAgreementInvoiceRowCreateUpdateSerializer(many=True)
+    payments = LandUseAgreementInvoicePaymentCreateUpdateSerializer(
+        many=True, required=False, allow_null=True
+    )
+    # Make total_amount, billed_amount, and type not required in the serializer and set them in create() if needed
+    total_amount = serializers.DecimalField(
+        max_digits=10, decimal_places=2, required=False
+    )
+    billed_amount = serializers.DecimalField(
+        max_digits=10, decimal_places=2, required=False
+    )
+    type = EnumField(enum=InvoiceType, required=False)
+
+    def override_permission_check_field_name(self, field_name):
+        # if field_name == "tenant":
+        #    return "recipient"
+
+        return field_name
+
+    def validate(self, attrs):
+        if not bool(attrs.get("recipient")) ^ bool(attrs.get("tenant")):
+            raise ValidationError(_("Either recipient or tenant is required."))
+
+        if (
+            attrs.get("tenant")
+            and attrs.get("tenant") not in attrs.get("lease").tenants.all()
+        ):
+            raise ValidationError(_("Tenant not found in lease"))
+
+        return attrs
+
+    def create(self, validated_data):
+        validated_data["state"] = InvoiceState.OPEN
+
+        if not validated_data.get("total_amount"):
+            total_amount = Decimal(0)
+            for row in validated_data.get("rows", []):
+                total_amount += row.get("amount", Decimal(0))
+
+            validated_data["total_amount"] = total_amount
+
+        if not validated_data.get("billed_amount"):
+            billed_amount = Decimal(0)
+            for row in validated_data.get("rows", []):
+                billed_amount += row.get("amount", Decimal(0))
+
+            validated_data["billed_amount"] = billed_amount
+
+        if not validated_data.get("type"):
+            validated_data["type"] = InvoiceType.CHARGE
+
+        # if validated_data.get("tenant"):
```

Comment:
Can the  commented part be removed or is this still under work?

Reply:
Yes, it can be removed. It is there still that I can compare to MVJ's invoice code. I will clean up these later.

Reply:
👍